`flat_map` doesn't play well with `count`

### Please complete the following tasks

- [X] I have searched the [discussions](https://github.com/winnow-rs/winnow/discussions)
- [X] I have searched the [open](https://github.com/winnow-rs/winnow/issues) and [rejected](https://github.com/winnow-rs/winnow/issues?q=is%3Aissue+label%3AS-wont-fix+is%3Aclosed) issues

### rust version

rustc 1.69.0-nightly (246eae2fa 2023-02-21)

### winnow version

0.3.0

### Minimal reproducible code

```rust
fn parse1(input: Partial<&IBytes>) -> IResult<Partial<&IBytes>, Vec<u16>> {
    let (input, num) = u8.parse_next(input)?;
    let (input, roads) = count(u16(Endianness::Big), num as usize).parse_next(input)?;

    Ok((input, roads))
}

fn parse2(input: Partial<&IBytes>) -> IResult<Partial<&IBytes>, Vec<u16>> {
    u8.flat_map(|num| count(u16(Endianness::Big), num as usize))
        .parse_next(input)
}
```


### Steps to reproduce the bug with the above code

`cargo check`

### Actual Behaviour

Doesn't compile

### Expected Behaviour

Compile as normal

### Additional Context

```
error[E0599]: no method named `parse_next` found for struct `winnow::combinator::FlatMap` in the current scope
  --> src/message.rs:96:10
   |
96 |         .parse_next(input)
   |          ^^^^^^^^^^ method not found in `FlatMap<fn(_) -> Result<(_, u8), ErrMode<_>> {u8::<_, _>}, [closure@message.rs:95:17], u8>`
```

From the doc, these two functions should be semantically the same. But the latter doesn't compile. Using `flat_map` with `take` is fine (as written in the doc).

Further investigation showed that `Parser` is not implemented for that type, so some trait bounds are not valid.

--- a/src/combinator/mod.rs
+++ b/src/combinator/mod.rs
@@ -605,7 +605,7 @@ impl<F, G, O1> FlatMap<F, G, O1> {
     }
 }
 
-impl<I, O1, O2, E, F: Parser<I, O1, E>, G: Fn(O1) -> H, H: Parser<I, O2, E>> Parser<I, O2, E>
+impl<I, O1, O2, E, F: Parser<I, O1, E>, G: FnMut(O1) -> H, H: Parser<I, O2, E>> Parser<I, O2, E>
     for FlatMap<F, G, O1>
 {
     fn parse_next(&mut self, i: I) -> IResult<I, O2, E> {

--- a/src/combinator/tests.rs
+++ b/src/combinator/tests.rs
@@ -6,7 +6,10 @@ use crate::error::Error;
 use crate::error::ErrorKind;
 use crate::error::Needed;
 use crate::error::ParseError;
+use crate::multi::count;
+use crate::number::u16;
 use crate::number::u8;
+use crate::number::Endianness;
 use crate::IResult;
 use crate::Parser;
 use crate::Partial;
@@ -115,6 +118,12 @@ fn test_parser_flat_map() {
         u8.flat_map(take).parse_next(input),
         Ok((&[103, 104][..], &[100, 101, 102][..]))
     );
+}
+
+#[allow(dead_code)]
+fn test_closure_compiles_195(input: &[u8]) -> IResult<&[u8], ()> {
+    u8.flat_map(|num| count(u16(Endianness::Big), num as usize))
+        .parse_next(input)
 }
 
 #[test]

